src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SslStreamSecurityUpgradeProvider.cs

```diff
@@ -566,16 +414,210 @@ internal override void Close(TimeSpan timeout)
+                rtStreamSocket.Control.IgnorableServerCertificateErrors.Add(
+                    ChainValidationResult.RevocationInformationMissing);
+                rtStreamSocket.Control.IgnorableServerCertificateErrors.Add(ChainValidationResult.RevocationFailure);
+
+                // SocketStream doesn't take a bitwise field of accepted protocols, but instead accepts a value specifying the highest
+                // protocol that can be negotiated. A check is made for each of the protocols in order to see if they've been requested
+                // by the binding and set the protection level to the UWP equivalent. This will have the effect of protectionLevel being
+                // set to the most secure protocol that was specified by client code. After the connection is established, if a protocol
+                // was negotiated which the binding didn't request, the connection needs to be aborted. This could happen for example if
+                // the requested SslProtocols was SslProtocols.Tls11 | SslProtocols.Tls12 and the server only supported SSL3 | Tls10. In
+                // this case, SocketProtectionLevel would be set to SocketProtectionLevel.Tls12, which would mean Tls10, Tls11 and Tls12
+                // are all acceptable protocols to negotiate. As the server is offering SSL3 | Tls10, the connection would be successfully
+                // negotiated using Tls10 which isn't allowed according to the binding configuration.
+                SocketProtectionLevel protectionLevel = SocketProtectionLevel.PlainSocket;
+                if ((_parent.SslProtocols & SslProtocols.Tls) != SslProtocols.None)
+                    protectionLevel = SocketProtectionLevel.Tls10;
+                if ((_parent.SslProtocols & SslProtocols.Tls11) != SslProtocols.None)
+                    protectionLevel = SocketProtectionLevel.Tls11;
+                if ((_parent.SslProtocols & SslProtocols.Tls12) != SslProtocols.None)
+                    protectionLevel = SocketProtectionLevel.Tls12;
+
+                // With SslStream, the hostname provided in the server certificate is provided to the client and verified in the callback.
+                // With UWP StreamSocket, the hostname needs to be provided to the call to UpgradeToSslAsync. The code to fetch the identity
+                // lives in the callback for CoreClr but needs to be pulled into this method for UWP.
+                EndpointAddress remoteAddress = RemoteAddress;
+                if (remoteAddress.Identity == null && remoteAddress.Uri != Via)
+                {
+                    remoteAddress = new EndpointAddress(Via);
+                }
+                EndpointIdentity identity;
+                if (!_parent.IdentityVerifier.TryGetIdentity(remoteAddress, out identity))
+                {
+                    SecurityTraceRecordHelper.TraceIdentityVerificationFailure(identity, null, GetType());
```

Comment:
Nit: could used named parameter pattern here so we know what these parameters mean (ex: the null).  Not blocking.